### ISSUE TITLE
Question: is there interest in adding warnings about questionable expressions?

### ISSUE THREAD

**Issue by UncombedCoconut** [lunarmodules/luacheck#43]
Hi, I've just fixed a bug in a game's Lua code, which looked like this:
```lua
if not mem.state==2 or not cargo_space then  -- expecting (not (mem.state == 2)) precedence
```
This is a rather elementary mistake, but after looking at other languages I think there could be a useful warning in here: Python reverses the precedence between "not" and comparisons. For C, which doesn't, gcc and Clang offer this warning option: "-Wlogical-not-parentheses: Warn about logical not used on the left hand side operand of a comparison".

As far as I can tell, Luacheck can parse expressions but doesn't have any warning category for valid-but-potentially-suspicious ones. Is that a use case which sounds interesting to support?

Thanks for your thoughts.

**Comment by arichard4** [comment]
I think that, rather than warn about logical not on the left hand side, we should just always warn for not on either side of an (in)equality, even if there's parentheses.

In particular, even the expected version above is suboptimal- this would be better: "if mem.state ~= 2"

> doesn't have any warning category for valid-but-potentially-suspicious ones

Luacheck doesn't have different levels of errors/warnings/issues. If you mean that this should be a new error type, yeah, that's straightforward to add.

**Comment by UncombedCoconut** [comment]
Thanks for the reply.
Two points of agreement: the real-life fix to my real-life example was to use ~=, and I did just mean this should (potentially) be a new error type.

It's hard to say how broad the warning should be.
The below example is perhaps a bad idea, but one whose hypothetical author probably meant what they wrote:
```lua
function xor(x, y)
    return x == not y
    -- Or a nil-tolerant "(not x) ~= (not y)"
end
```
So any choice of rule would be making a tradeoff. Your version sounds reasonable: over-strict if a user just wants to catch bugs, but probably accurate for catching code that could be written better.

**Comment by arichard4** [comment]
That's a good example case!

**Comment by arichard4** [comment]
I've pushed a proposed change, above. It has separate warning categories for not (x == y) and not x == y; it also only looks at the left-hand side for both.

There's still one case where it produces a false positive:

```
local metatable = {
    __le = function(a,b) --[[ Return something --]] end,
    __gt = function(a,b) return not (b <= a) end
}
```

Thoughts?

**Comment by UncombedCoconut** [comment]
I disappeared around the holidays and forgot to mention: I like this. I think it's fine to add the warning about `not (x OP y)` by default. Legitimate use cases should be exceptional and OK to handle with inline options. Even after disabling 581, users would get the more conservative warning I first asked about.
